packages/ddp-client/client/queueStubsHelpers.js

```diff
@@ -20,7 +20,7 @@ export const loadAsyncStubHelpers = () => {
 
     queue = queue.finally(() => {
       fn(resolve, reject);
-      return promise;
+      return promise.stubPromise;
```

Comment:
This works since all except the one `fn` are sync, but it seems unexpected for it to not wait on `promise`. It might be better and simpler for `applyAsync` to handle creating its own promise to return, and `queueFunction` can avoid having to deal with `stubPromise`/`serverPromise`.

Reply:
I'm getting familiar with this code. Attempted to refactor this but breaking things along the way. I may retry to refactor in another PR while this PR focus on what tries to fix, which is done and tests passes green now.

Reply:
Or maybe I am not understanding well your suggestions here. Could you guide me?

Reply:
I haven't tested this code, but something like this in the `applyAsync` wrapper:

```js
    let applyAsyncResolve;
    let applyAsyncReject;
    let stubPromiseResolver;
    let serverPromiseResolver;

    const applyAsyncPromise = new Promise((resolve, reject) => {
      applyAsyncResolve = resolve;
      applyAsyncReject = reject;
    });
    const stubPromise = new Promise((r) => (stubPromiseResolver = r));
    const serverPromise = new Promise((r) => (serverPromiseResolver = r));

    queueFunction(
      (next) => {
        let hasStub = false;
        let finished = false;

        Meteor._setImmediate(() => {
          const promise = oldApplyAsync.apply(this, args);
          stubPromiseResolver(promise.stubPromise);
          serverPromiseResolver(promise.serverPromise);
          hasStub = !!promise.stubPromise;

          if (hasStub) {
            promise.stubPromise.finally(() => {
              finished = true;
              next();
            });
          } else {
            next();
          }

          promise
            .then((result) => {
              applyAsyncResolve(result);
            })
            .catch((err) => {
              applyAsyncReject(err);
            });
        });

        Meteor._setImmediate(() => {
          if (hasStub && !finished) {
            console.warn(
              `Method stub (${name}) took too long and could cause unexpected problems. Learn more at https://github.com/zodern/fix-async-stubs/#limitations`
            );
          }
        });
      }
    );

    return applyAsyncPromise;
```

Basically, instead of returning the promise from queueFunction, have applyAsync create a separate promise that is returned. This fixes the api for queueFunction by allowing the resolve function given to the callback to be used to control the queue.

Originally, applyAsync used the same promise as queueFunction since Meteor 2 used a promise for the stub and a callback for the server result. However, now that Meteor 3 uses the returned function for the server result, reusing the promise doesn't make as much sense anymore.